extmod/modlwip.c

```diff
@@ -0,0 +1,974 @@
+    ECONNRESET,    /* ERR_RST        -11     Connection reset.        */
+    ENOTCONN,      /* ERR_CLSD       -12     Connection closed.       */
+    ENOTCONN,      /* ERR_CONN       -13     Not connected.           */
+    EIO,           /* ERR_ARG        -14     Illegal argument.        */
+    -1,            /* ERR_IF         -15     Low-level netif error    */
+    EBADF,         /* Not an ERR     -16     Closed socket (null pcb) */
+};
+
+/*******************************************************************************/
+// The socket object provided by lwip.socket.
+
+#define MOD_NETWORK_AF_INET (2)
+#define MOD_NETWORK_AF_INET6 (10)
+
+#define MOD_NETWORK_SOCK_STREAM (1)
+#define MOD_NETWORK_SOCK_DGRAM (2)
+#define MOD_NETWORK_SOCK_RAW (3)
+
+typedef struct _lwip_socket_obj_t {
+    mp_obj_base_t base;
+
+    void *pcb;
+    void *incoming;
+    byte peer[4];
+    mp_uint_t peer_port;
+    mp_uint_t timeout;
+    uint16_t leftover_count;
+
+    uint8_t domain;
+    uint8_t type;
+
+    // 0 = unconnected, 1 = connecting, 2 = connected, 3 = other side closed
+    int8_t connected;
+} lwip_socket_obj_t;
+
+/*******************************************************************************/
+// Callback functions for the lwIP raw API.
+
+// Callback for incoming UDP packets. We simply stash the packet and the source address,
+// in case we need it for recvfrom.
+STATIC void _lwip_udp_incoming(void *arg, struct udp_pcb *upcb, struct pbuf *p, ip_addr_t *addr, u16_t port) {
+    lwip_socket_obj_t *socket = (lwip_socket_obj_t *)arg;
+
+    if (socket->incoming != NULL) {
+        // That's why they call it "unreliable". No room in the inn, drop the packet.
+        pbuf_free(p);
+    } else {
+        socket->incoming = (void *)p;
+        socket->peer_port = (mp_uint_t)port;
+        memcpy(&(socket->peer), addr, 4);
+    }
+}
+
+// Callback for general tcp errors.
+STATIC void _lwip_tcp_error(void *arg, err_t err) {
+    lwip_socket_obj_t *socket = (lwip_socket_obj_t *)arg;
+
+    // Pass the error code back via the connection variable.
+    socket->connected = err;
+    // If we got here, the lwIP stack either has deallocated or will deallocate the pcb.
+    socket->pcb = NULL;
+}
+
+// Callback for tcp connection requests. Error code err is unused. (See tcp.h)
+STATIC err_t _lwip_tcp_connected(void *arg, struct tcp_pcb *tpcb, err_t err) {
+    lwip_socket_obj_t *socket = (lwip_socket_obj_t *)arg;
+
+    socket->connected = 2;
+    return ERR_OK;
+}
+
+// Callback for incoming tcp connections.
+STATIC err_t _lwip_tcp_accept(void *arg, struct tcp_pcb *newpcb, err_t err) {
+    lwip_socket_obj_t *socket = (lwip_socket_obj_t *)arg;
+
+    if (socket->incoming != NULL) {
+        // We need to handle this better. This single-level structure makes the
+        // backlog setting kind of pointless. FIXME
+        return ERR_BUF;
+    } else {
+        socket->incoming = (void *)newpcb;
```

Comment:
@galenhz : So, here very suspicious typesafety problem, masked by void*: socket->incoming is generally pbuf, but here you store pcb to it, and that's very suspicious, because there's already socket->pcb. What's that? Optimization? But then it should be documented by both words, and of proper unions.